pandas/core/extensions.py

```diff
@@ -0,0 +1,216 @@
+
+        Raises
+        ------
+        TypeError
+
+        Notes
+        -----
+        The default implementation checks if 'string' matches your
+        type's name. If so, it calls your class with no arguments.
+        """
+        if string == cls.name:
+            return cls()
+        else:
+            raise TypeError("Cannot construct a '{}' from "
+                            "'{}'".format(cls, string))
+
+    @classmethod
+    def is_dtype(cls, dtype: T.Union[str, np.dtype]) -> bool:
+        """Check if we match 'dtype'
+
+        Parameters
+        ----------
+        dtype : str or dtype
+
+        Returns
+        -------
+        is_dtype : bool
+
+        Notes
+        -----
+        The default implementation is True if
+
+        1. 'dtype' is a string that returns true for
+           ``cls.construct_from_string``
+        2. 'dtype' is ``cls`` or a subclass of ``cls``.
+        """
+        if isinstance(dtype, str):
+            try:
+                return isinstance(cls.construct_from_string(dtype), cls)
+            except TypeError:
+                return False
+        else:
+            return issubclass(dtype, cls)
+
+
+class ExtensionArray(metaclass=abc.ABCMeta):
+    """Abstract base class for custom array types
+
+    pandas will recognize instances of this class as proper arrays
+    with a custom type and will not attempt to coerce them to objects.
+
+    Subclasses are expected to implement the following methods.
+    """
+    # ------------------------------------------------------------------------
+    # Must be a Sequence
+    # ------------------------------------------------------------------------
+    @abc.abstractmethod
+    def __iter__(self):
+        pass
+
+    @abc.abstractmethod
+    def __len__(self):
+        pass
+
+    # ------------------------------------------------------------------------
+    # Required attributes
+    # ------------------------------------------------------------------------
+    @property
+    @abc.abstractmethod
+    def dtype(self):
+        pass
+
+    @property
+    @abc.abstractmethod
+    def shape(self) -> T.Tuple[int, ...]:
+        pass
+
+    @property
+    @abc.abstractmethod
+    def ndim(self) -> int:
+        pass
+
+    @property
+    @abc.abstractmethod
+    def nbytes(self) -> int:
+        # TODO: default impl?
+        pass
+
+    # ------------------------------------------------------------------------
+    # Additional Methods
+    # ------------------------------------------------------------------------
+    @abc.abstractmethod
+    def isna(self) -> T.Sequence[bool]:
+        # TODO: narrow this type?
+        pass
+
+    # ------------------------------------------------------------------------
+    # Indexing methods
+    # ------------------------------------------------------------------------
+    @abc.abstractmethod
+    def take(self, indexer, allow_fill=True, fill_value=None):
+        """For slicing"""
```

Comment:
It would be good to clarify if this is the NumPy version of `take`, or the pandas internal version that uses `-1` to indicate fill values.

If it's the later (as I suspect is the case) it could be valuable to expose some of our internal utility functions for this rather than force users to re-implement it. It is non-trivial to get right in a performant way.